docs/design/coreclr/botr/data_contracts_for_coreclr.md

```diff
@@ -0,0 +1,430 @@
+    {
+        // Get contract from the predefined contract database
+        static class PredefinedContracts
+        {
+            public static IContract GetContract(string name, uint version, Target target)
+            {
+                // Do some lookup and allocate an instance of the contract requested
+            }
+        }
+
+        class MethodTableContract_1 : IAlgorithmContract
+        {
+            Target _target;
+            FieldLayout _dynamicTypeIdField;
+            FieldLayout _baseSizeField;
+
+            public MethodTableContract_1(Target target) { _target = target; }
+
+            void IAlgorithmContract.Init()
+            {
+                _dynamicTypeIdField = _target.Contract.GetFieldLayout("MethodTable", "dynamicTypeId");
+                if (_dynamicTypeIdField.Type != FieldType.Int32Type)
+                    throw new InvalidContractSpec();
+                _baseSizeField = _target.Contract.GetFieldLayout("MethodTable", "baseSize");
+                if (_baseSizeField.Type != FieldType.Int32Type)
+                    throw new InvalidContractSpec();
+            }
+
+            public virtual int DynamicTypeID(TargetPointer methodTablePointer) { return _target.ReadInt32(_target.GetTargetPointerForField(methodTablePointer, _dynamicTyepIdField)); }
+            public virtual int BaseSize(TargetPointer methodTablePointer) { return _target.ReadInt32(_target.GetTargetPointerForField(methodTablePointer, _baseSizeField)); }
+        }
+
+        // This is used for version 2 of the contract, where the dynamic type id is no longer present
+        class MethodTableContract_2 : IAlgorithmContract
+        {
+            Target _target;
+            FieldLayout _baseSizeField;
+
+            public MethodTableContract_2(Target target) { _target = target; }
+
+            void IAlgorithmContract.Init()
+            {
+                _baseSizeField = _target.Contract.GetFieldLayout("MethodTable", "baseSize");
+                if (_baseSizeField.Type != FieldType.Int32Type)
+                    throw new InvalidContractSpec();
+            }
+
+            public virtual int DynamicTypeID(TargetPointer methodTablePointer) { throw new NotImplementedException(); }
+
+            public virtual int BaseSize(TargetPointer methodTablePointer) { return _target.ReadInt32(_target.GetTargetPointerForField(methodTablePointer, _baseSizeField)); }
+        }
+    }
+}
+```
+
+Within the contract algorithm specification, the intention is that all interesting api work is done by using an instance of the `Target` class.
+
+## Arrangement of contract specifications in the repo
+
+Specs shall be stored in the repo in a set of directories. `docs/design/datacontracts`
+
+### Global Value Contracts
+For global value contracts they shall be stored in `docs/design/datacontracts/globals/<contract_name>/<version>`
+
+The name of each contract shall be `docs/design/datacontracts/globals/<contract_name>/<contract_name>_<version>.md`
```

Comment:
I actually would like to see machine generation/handling of the data structure stuff, so that the runtime level source code actually only works with offsets, and yet the stream could output only the more optimized type version data. As I said in another comment, I think this deserves a bit of experimentation to see how gnarly it is.